File: frontend/src/services/transformerSizing.ts
import { z } from "zod";

import { calculationRunSummarySchema } from "./calculationRun";
import { jurisdictionProfileSchema } from "./cableContract";
import { exactDecimalSchema } from "./faultContract";
import { ApiError } from "./http";
import { loadScenarioSchema } from "./loadDemandContract";
import {
  DEMAND_POWER_FACTOR_REQUIRED,
  transformerRedundancyModeSchema,
  transformerSizingStatusSchema,
  transformerSizingWarningCodeSchema,
} from "./transformerSizingContract";

const optionalRequestText = (maxLength: number) =>
  z.string().trim().max(maxLength).nullable().optional();

// A factor left out means NOT ESTABLISHED, never 1: the engine sizes with 1,
// names the factor in a warning and reports REVIEW_REQUIRED (Master Prompt
// A16 (a)). The form must omit a blank field instead of sending "1".
const optionalFactor = exactDecimalSchema.optional();

export const transformerSizingRequestSchema = z
  .object({
    code: z.string().trim().min(1).max(50),
    name: z.string().trim().min(1).max(200),
    demand_power_kw: exactDecimalSchema,
    // An engineering input, not a margin: it has no default and is required.
    demand_power_factor: exactDecimalSchema.optional(),
    available_unit_ratings_kva: z.array(exactDecimalSchema).min(1),
    future_growth_factor: optionalFactor,
    design_margin_factor: optionalFactor,
    ambient_derating_factor: optionalFactor,
    altitude_derating_factor: optionalFactor,
    harmonic_derating_factor: optionalFactor,
    duty_units: z.number().int().positive().optional(),
    standby_units: z.number().int().nonnegative().optional(),
    redundancy_mode: transformerRedundancyModeSchema.optional(),
    scenario: loadScenarioSchema.optional(),
    jurisdiction_profile: jurisdictionProfileSchema.optional(),
    notes: optionalRequestText(1000),
  })
  .strict()
  .superRefine((values, ctx) => {
    if (values.demand_power_factor === undefined) {
      // Reported as a missing value rather than a custom rule, so the form's
      // describeValidationIssue turns it into "... is required." instead of
      // pasting the raw sentence after the label.
      ctx.addIssue({
        code: "invalid_type",
        expected: "string",
        received: "undefined",
        message: DEMAND_POWER_FACTOR_REQUIRED,
        path: ["demand_power_factor"],
      });
    }
  });

export const transformerSizingWarningSchema = z
  .object({
    code: transformerSizingWarningCodeSchema,
    message: z.string(),
  })
  .strict();

export const transformerSizingResponseSchema = z
  .object({
    code: z.string(),
    name: z.string(),
    scenario: loadScenarioSchema,
    redundancy_mode: transformerRedundancyModeSchema,
    demand_power_kw: z.string(),
    demand_power_factor: z.string(),
    base_demand_kva: z.string(),
    // The factor the sizing used; "1" when it was not established.
    future_growth_factor: z.string(),
    future_demand_kva: z.string(),
    design_margin_factor: z.string(),
    design_required_kva: z.string(),
    combined_derating_factor: z.string(),
    required_nameplate_capacity_kva: z.string(),
    duty_units: z.number().int(),
    standby_units: z.number().int(),
    total_units: z.number().int(),
    required_unit_rating_kva: z.string(),
    // Null together with the four below when no rating was adequate (NO_SOLUTION).
    selected_unit_rating_kva: z.string().nullable(),
    installed_nameplate_capacity_kva: z.string().nullable(),
    derated_duty_capacity_kva: z.string().nullable(),
    spare_derated_capacity_kva: z.string().nullable(),
    loading_percent: z.string().nullable(),
    status: transformerSizingStatusSchema,
    warnings: z.array(transformerSizingWarningSchema),
    jurisdiction_profile: jurisdictionProfileSchema,
  })
  .strict();

export const transformerRunCreateRequestSchema = z
  .object({
    study: transformerSizingRequestSchema,
    project_revision_id: z.string().uuid().optional(),
    notes: optionalRequestText(2000),
  })
  .strict();

export const transformerRunResponseSchema = z
  .object({
    run: calculationRunSummarySchema,
    result: transformerSizingResponseSchema,
  })
  .strict();

const errorDetailSchema = z.union([
  z.string(),
  z.array(
    z
      .object({
        loc: z.array(z.union([z.string(), z.number()])),
        msg: z.string(),
        type: z.string(),
      })
      .passthrough(),
  ),
]);

const errorResponseSchema = z.object({
  detail: errorDetailSchema,
});

export type TransformerSizingRequest = z.infer<typeof transformerSizingRequestSchema>;
export type TransformerSizingWarning = z.infer<typeof transformerSizingWarningSchema>;
export type TransformerSizingResponse = z.infer<typeof transformerSizingResponseSchema>;
export type TransformerRunCreateRequest = z.infer<typeof transformerRunCreateRequestSchema>;
export type TransformerRunResponse = z.infer<typeof transformerRunResponseSchema>;

function formatApiError(data: unknown, status: number): string {
  const parsed = errorResponseSchema.safeParse(data);

  if (!parsed.success) {
    return `Transformer sizing failed (HTTP ${status}).`;
  }

  if (typeof parsed.data.detail === "string") {
    return parsed.data.detail;
  }

  return parsed.data.detail
    .map((item) => `${item.loc.join(".")}: ${item.msg}`)
    .join("; ");
}

/**
 * Calculate a transformer sizing study and persist it as a run revision.
 *
 * The request carries the whole run body, because a study's notes belong to the
 * run, not to the sizing the engine calculates. The returned summary carries the
 * run ID, engine version, profile, reference status and content hash shown in
 * the traceability panel.
 *
 * With a project revision id the run is stored under that revision of the selected
 * project; without one it belongs to no project. An unchanged study returns the
 * stored revision with HTTP 200 instead of a new one (A11); both are valid here.
 */
export async function createTransformerRun(
  payload: TransformerRunCreateRequest,
  signal?: AbortSignal,
  projectRevisionId?: string,
): Promise<TransformerRunResponse> {
  const validatedPayload = transformerRunCreateRequestSchema.parse(payload);
  const timeoutSignal = AbortSignal.timeout(30_000);
  const requestSignal = signal
    ? AbortSignal.any([signal, timeoutSignal])
    : timeoutSignal;

  const response = await fetch("/api/v1/electrical/transformer-sizing/runs", {
    method: "POST",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(
      projectRevisionId === undefined
        ? validatedPayload
        : { ...validatedPayload, project_revision_id: projectRevisionId },
    ),
    cache: "no-store",
    signal: requestSignal,
  });

  const data: unknown = await response.json().catch(() => null);

  if (!response.ok) {
    throw new ApiError(formatApiError(data, response.status), response.status);
  }

  const parsed = transformerRunResponseSchema.safeParse(data);

  if (!parsed.success) {
    throw new Error("Unexpected response from the KES Electrical OS transformer runs API.");
  }

  return parsed.data;
}

File: backend/app/schemas/transformer_sizing.py
"""
Pydantic schemas for transformer source-sizing calculations.
KESE-S2-M5
"""

from decimal import Decimal
from typing import Annotated, Self

from pydantic import (
    BaseModel,
    BeforeValidator,
    ConfigDict,
    Field,
    StrictInt,
    model_validator,
)

from app.domain.electrical.jurisdiction.jurisdiction_models import (
    JurisdictionProfile,
)
from app.domain.electrical.loads.models import LoadScenario
from app.domain.electrical.sources.models import (
    TransformerRedundancyMode,
    TransformerSizingInput,
)
from app.domain.electrical.sources.results import (
    TransformerSizingResult,
    TransformerSizingStatus,
    TransformerSizingWarningCode,
)


def _reject_float(value: object) -> object:
    """Reject binary floating-point engineering inputs."""

    if isinstance(value, float):
        raise ValueError(
            "engineering decimal values must be provided as strings, integers, or Decimal values"
        )

    return value


type ExactDecimal = Annotated[
    Decimal,
    BeforeValidator(_reject_float),
]

type PositiveExactDecimal = Annotated[
    Decimal,
    BeforeValidator(_reject_float),
    Field(
        gt=Decimal("0"),
        max_digits=38,
        decimal_places=18,
    ),
]

type RatioExactDecimal = Annotated[
    Decimal,
    BeforeValidator(_reject_float),
    Field(
        gt=Decimal("0"),
        le=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    ),
]

type FactorExactDecimal = Annotated[
    Decimal,
    BeforeValidator(_reject_float),
    Field(
        ge=Decimal("1"),
        max_digits=38,
        decimal_places=18,
    ),
]


class _RequestBase(BaseModel):
    """Shared transformer-sizing request configuration."""

    model_config = ConfigDict(
        extra="forbid",
        str_strip_whitespace=True,
    )


class _ResponseBase(BaseModel):
    """Shared transformer-sizing response configuration."""

    model_config = ConfigDict(
        from_attributes=True,
        extra="forbid",
    )


class TransformerSizingRequest(_RequestBase):
    """API request for transformer source sizing."""

    code: str = Field(
        min_length=1,
        max_length=50,
    )
    name: str = Field(
        min_length=1,
        max_length=200,
    )

    demand_power_kw: PositiveExactDecimal
    demand_power_factor: RatioExactDecimal

    available_unit_ratings_kva: tuple[
        PositiveExactDecimal,
        ...,
    ] = Field(
        min_length=1,
    )

    # A blank factor means NOT ESTABLISHED and reaches the engine as None: it sizes
    # with 1, names the factor in a warning and reports REVIEW_REQUIRED (A16 (a)).
    # The schema never fills a default in its place.
    future_growth_factor: FactorExactDecimal | None = None
    design_margin_factor: FactorExactDecimal | None = None

    ambient_derating_factor: RatioExactDecimal | None = None
    altitude_derating_factor: RatioExactDecimal | None = None
    harmonic_derating_factor: RatioExactDecimal | None = None

    duty_units: StrictInt = Field(
        default=1,
        gt=0,
    )
    standby_units: StrictInt = Field(
        default=0,
        ge=0,
    )

    redundancy_mode: TransformerRedundancyMode = TransformerRedundancyMode.NONE
    scenario: LoadScenario = LoadScenario.NORMAL

    # The sizing itself has no profile-dependent value; the profile is carried for
    # the result, the run record and the design basis of the project revision.
    jurisdiction_profile: JurisdictionProfile = JurisdictionProfile.IN

    notes: str | None = None

    @model_validator(mode="after")
    def validate_rating_schedule_and_redundancy(
        self,
    ) -> Self:
        """Validate controlled ratings and redundancy arrangement."""

        if len(self.available_unit_ratings_kva) != len(set(self.available_unit_ratings_kva)):
            raise ValueError("available transformer ratings must be unique")

        if self.available_unit_ratings_kva != tuple(sorted(self.available_unit_ratings_kva)):
            raise ValueError("available transformer ratings must be in ascending order")

        if self.redundancy_mode is TransformerRedundancyMode.NONE and self.standby_units != 0:
            raise ValueError("NONE redundancy requires standby_units to be 0")

        if self.redundancy_mode is TransformerRedundancyMode.N_PLUS_1 and self.standby_units != 1:
            raise ValueError("N_PLUS_1 redundancy requires exactly one standby unit")

        if (
            self.redundancy_mode is TransformerRedundancyMode.TWO_N
            and self.standby_units != self.duty_units
        ):
            raise ValueError("TWO_N redundancy requires standby_units to equal duty_units")

        return self

    def to_domain(self) -> TransformerSizingInput:
        """Convert the validated API request to a domain record."""

        return TransformerSizingInput(
            code=self.code,
            name=self.name,
            demand_power_kw=self.demand_power_kw,
            demand_power_factor=self.demand_power_factor,
            available_unit_ratings_kva=(self.available_unit_ratings_kva),
            future_growth_factor=self.future_growth_factor,
            design_margin_factor=self.design_margin_factor,
            ambient_derating_factor=(self.ambient_derating_factor),
            altitude_derating_factor=(self.altitude_derating_factor),
            harmonic_derating_factor=(self.harmonic_derating_factor),
            duty_units=self.duty_units,
            standby_units=self.standby_units,
            redundancy_mode=self.redundancy_mode,
            scenario=self.scenario,
            notes=self.notes,
        )


class TransformerSizingWarningResponse(_ResponseBase):
    """Structured warning returned by the sizing API."""

    code: TransformerSizingWarningCode
    message: str


class TransformerSizingResponse(_ResponseBase):
    """API response for a transformer-sizing calculation."""

    code: str
    name: str

    scenario: LoadScenario
    redundancy_mode: TransformerRedundancyMode

    demand_power_kw: Decimal
    demand_power_factor: Decimal
    base_demand_kva: Decimal

    future_growth_factor: Decimal
    future_demand_kva: Decimal

    design_margin_factor: Decimal
    design_required_kva: Decimal

    combined_derating_factor: Decimal
    required_nameplate_capacity_kva: Decimal

    duty_units: int
    standby_units: int
    total_units: int

    required_unit_rating_kva: Decimal
    selected_unit_rating_kva: Decimal | None

    installed_nameplate_capacity_kva: Decimal | None
    derated_duty_capacity_kva: Decimal | None
    spare_derated_capacity_kva: Decimal | None
    loading_percent: Decimal | None

    status: TransformerSizingStatus

    warnings: tuple[
        TransformerSizingWarningResponse,
        ...,
    ]

    jurisdiction_profile: JurisdictionProfile

    @classmethod
    def from_domain(
        cls,
        result: TransformerSizingResult,
        *,
        jurisdiction_profile: JurisdictionProfile,
    ) -> Self:
        """
        Create the response from the domain result and the requested profile.

        The sizing engine carries no jurisdiction profile, so the profile of the
        request is echoed here rather than read from the result.
        """

        return cls(
            code=result.code,
            name=result.name,
            scenario=result.scenario,
            redundancy_mode=result.redundancy_mode,
            demand_power_kw=result.demand_power_kw,
            demand_power_factor=result.demand_power_factor,
            base_demand_kva=result.base_demand_kva,
            future_growth_factor=result.future_growth_factor,
            future_demand_kva=result.future_demand_kva,
            design_margin_factor=result.design_margin_factor,
            design_required_kva=result.design_required_kva,
            combined_derating_factor=result.combined_derating_factor,
            required_nameplate_capacity_kva=(result.required_nameplate_capacity_kva),
            duty_units=result.duty_units,
            standby_units=result.standby_units,
            total_units=result.total_units,
            required_unit_rating_kva=result.required_unit_rating_kva,
            selected_unit_rating_kva=result.selected_unit_rating_kva,
            installed_nameplate_capacity_kva=(result.installed_nameplate_capacity_kva),
            derated_duty_capacity_kva=result.derated_duty_capacity_kva,
            spare_derated_capacity_kva=result.spare_derated_capacity_kva,
            loading_percent=result.loading_percent,
            status=result.status,
            warnings=tuple(
                TransformerSizingWarningResponse.model_validate(warning)
                for warning in result.warnings
            ),
            jurisdiction_profile=jurisdiction_profile,
        )


__all__ = [
    "TransformerSizingRequest",
    "TransformerSizingResponse",
    "TransformerSizingWarningResponse",
]
